### ISSUE TITLE
Developer API

### ISSUE THREAD

**Issue by flowww-dev** [soflyy/breakdance-bugs#730]
It would be nice to have a simple Developer API like Typerocket to create Custom Elements and Custom Fieldtypes.
Also a Utility Class could help to validate and specify $propertiesData in %%SSR%%.

So instead of  use:
$id = $propertiesData['content']['section']['field'];

The Utility Class could be used like:

$id = $this->getAsInt( 'content/section', 'id', 0 );
The last parameter is a default Value;

Or:

$text =  $this->getAsString( 'content/section', 'text', 'Hello World', $PREPEND_IF_HAS_VALUE, $APPEND_IF_HAS_VALUE );

Or:

$isVisible = $this->getAsBool( 'design/section', 'isVisible', true );

The API could look like (for example a simple Video Element):

BDC::add_element('video', 'Video', 'Medien', 'ICON')
    ->register_style( 'video', BD_PUBLIC_VC_URI.'/video.css' )
    ->register_script( 'video', BD_PUBLIC_VC_URI.'/video.js')
    ->register_style( 'vendor-plyr', BD_PUBLIC_VENDOR_URI.'/plyr/dist/plyr.css')
    ->register_script( vendor-plyr', BD_PUBLIC_ASSETS_URI.'/js/plyr/plyr.min.js')
    ->add_params( [
        BDF::add_group( 'Inhalt', [
            BDF::add_select( 'type', 'Playtype', [ '---' => '', 'Player' => 'player', 'Viewport Autoplay' => 'viewport-autoplay' ] ),
            BDF::add_file( 'mp4', 'MP4' ),
            BDF::add_image( 'poster', 'Image' ),
        ]),
    ])
    ->render(
        function() {
            /* @var $this VCU */
            $mp4 = '';
            $poster = '';
            $player_args = '';
            $type = $this->get_as_string( 'type', 'player' );
            if( $type == 'viewport-autoplay' ) {
                $player_args = ' muted playsinline autoplay loop preload="none"';
            }
            $mp4 = $this->get_as_file_url( 'mp4', $mp4 );
            $poster = $this->get_as_image_url( 'poster',  $poster );
            $this->render_container_start( null, null, [ 'data-type' => $type ] );
            ?>
            <div class="wrapper">
                <video class="video" id="player" preload="auto" <?php echo $player_args ?> poster="<?php echo $poster ?>" data-poster="<?php echo $poster ?>">
                    <source src="<?php echo $mp4 ?>" type="video/mp4" />
                </video>
            </div>
            <?php
            $this->render_container_end( );
        }
    );

**Comment by a-dubiel** [comment]
Did you try using Element Studio?

https://breakdance.com/documentation/developers/element-studio/

**Comment by flowww-dev** [comment]
Yes, but a API for Power Developer should be also great. Element Studio is really cool, but I think many Developer missing some features like transpiing, autocompletion, Copilot, SCSS (or other) npm and other stuff.

This can be only managed by an IDE and therefore a simple Framework has none of the other Page bUilders (the only one is WPBackery).

Greetings

Marius

**Comment by flowww-dev** [comment]
As I said. Element Studio is really great. But I am missing scss, webpack and so on - 
but these can also be my personal impressions.. Ican live with Element Studio ;-) . Factory classes sounds great. It would be also great, if it were possible to transfer the elements not just to a separate plugin. I mostly use my own themes. I just like having everything in one place. That's why it would be great if Element Studio also offered the option of transferring the created elements to your own theme and still having the option of expanding them in Element Studio.

**Comment by flowww-dev** [comment]
Is it possible to unregister Icons (FontAwesome and so on) and unregister Elements to hide some for Customers?

**Comment by flowww-dev** [comment]
Yes it works with anonymous functions ^^:

That's the only, I need to write to add a dynamic field.

```

WPB_Breakdance::register_dynamic_field( 'test', 'Test', 'Test' )
    ->set_handler(
        function( $attributes ) {
            return 'test';
        }
    );

```
